config/src/config/network_config.rs

```diff
@@ -153,6 +160,45 @@ impl NetworkConfig {
         .expect("peer id should be present")
     }
 
+    pub fn shared_val_netaddr_keys(&mut self) -> HashMap<netaddr::KeyVersion, netaddr::Key> {
+        // The shared validator network address key is only used in the
+        // validator network. For all other cases, we'll just return an empty key
+        // map.
+        if self.network_id != NetworkId::Validator {
+            return HashMap::new();
+        }
+
+        let keys = match &mut self.shared_val_netaddr_keys {
+            SharedValNetAddrKeys::FromConfig { keys } => keys.drain().collect(),
+            SharedValNetAddrKeys::FromStorage { backend } => {
+                let storage = Storage::from(&*backend);
+                let key_bytes = storage
+                    .get(libra_global_constants::SHARED_VAL_NETADDR_KEY)
+                    .expect("Unable to read shared_val_netaddr_keys")
+                    .value
+                    .bytes()
+                    .expect("Expected bytes for shared_val_netaddr_keys");
+                let key = netaddr::Key::try_from(key_bytes.as_slice())
+                    .expect("Expected shared_val_netaddr_keys bytes to be well-formatted");
+
+                // TODO(philiphayes): add rotation + multiple keys support. for
+                // now just assume a constant version.
+                let key_version = SHARED_VAL_NETADDR_KEY_VERSION;
+
+                let mut keys = HashMap::new();
+                keys.insert(key_version, key);
+                keys
+            }
+            SharedValNetAddrKeys::None => HashMap::new(),
+        };
+
+        assert!(
+            !keys.is_empty(),
+            "shared_val_netaddr_keys should not be empty for validators"
+        );
+        keys
+    }
+
```

Comment:
I wonder if this is a bit too future proof. Maybe for now we just want a single key?

Reply:
in my mind, it's just the secure storage model that's behind and will catch up after v1. though, it doesn't really matter to me either way.